### ISSUE TITLE
DockPanel LastChildFill doesn't consider visibility of control

### ISSUE THREAD

**Issue by trodent83** [dotnet/wpf#150]
The dockpanel doesn't checks if the last Control in it's list is Visible, Hiden, or Collapsed, and always reserves the remaining place for it.
This if the Control is collapsed doesn't really make sence.

It would be good if the place calculation of the panel would also consider the visibility property of the Control.

**Comment by thomasclaudiushuber** [comment]
Hi @trodent83, I just had a look at this. I think this is on purpose, as the property says that the last child should fill the available space. Let's just write down what it does today:

### LastChildFill == false

The DockPanel reacts correctly on the Visibility property. Change the Visibility on the first Button below to Hidden, and it will reserve space, with collapsed it doesn't reserve space.

```
<DockPanel LastChildFill="False">
        <Button Content="1" DockPanel.Dock="Top" Visibility="Collapsed"/>
        <Button Content="2" DockPanel.Dock="Top"/>
 </DockPanel>
```

### LastChildFill == true

If this is set to true, that means that the last child always fills the leftover space. That means setting the Visibility-Property on the last child to Collapsed will have exactly the same effect as setting it to Hidden, the elements is not visible, but it's still considered as the last element in the DockPanel.

```
<DockPanel LastChildFill="False">
        <Button Content="1" DockPanel.Dock="Top" />
        <Button Content="2" DockPanel.Dock="Top" Visibility="Collapsed"/>
 </DockPanel>
```


What you, @trodent83, mean is that the LastChildFill-property should react on the last child that is not collapsed. That means that in the DockPanel above the Button 2 should be ignored and Button 1 should fill the available space. To get the effect like we have it already today, a dev would have to set the Visibility property of the last child to Hidden instead of Collapsed. Setting it to Collapsed would have the same effect as removing that collapsed element:

```
<DockPanel LastChildFill="False">
        <Button Content="1" DockPanel.Dock="Top" />
       <!--  <Button Content="2" DockPanel.Dock="Top" Visibility="Collapsed"/> -->
 </DockPanel>
```

Yes, it kind of makes sense from the Visibility-property point of view, and it would allow something like the DockPanel below, where you either want to show "MainContent 1" or "MainContent 2" in the leftover space. But setting Visibility to Collapsed on MainContent 2 doesn't change anything, as MainContent 2 is still considered the last child:

```
  <DockPanel LastChildFill="True">
        <Button Content="Main menu" DockPanel.Dock="Top"/>
        <Button Content="MainContent 1" />
        <Button Content="MainContent 2" Visibility="Collapsed" />
    </DockPanel>
```

While I understand this issue, I'm not sure if it is a good idea to implement it, as it might break a lot of existing WPF code. If you want a scearnio like the one above, you can easily implement it like this:

```
    <DockPanel LastChildFill="True">
        <Button Content="Main menu" DockPanel.Dock="Top"/>
        <Grid>
            <Button Content="MainContent 1" />
            <Button Content="MainContent 2" Visibility="Collapsed" />
        </Grid>
    </DockPanel>
```

That's the reason why I appreciate this issue, but why I wouldn't change anything. What do you think, @trodent83?

**Comment by trodent83** [comment]
Hi @thomasclaudiushuber,

I think this was simply not testet and thought through.
I think it should not break anything because Setting something Explicitly to Collapsed is done to make it collapsed, but it is also possible to add an extra property to the dockpanel and switch between the legacy mode and the new mode.

That is basicly what I am doing in my Code Basis and because of that I know the effort to correct this and so have a normal solution in the Basis classes of the WPF is not too much.

**Comment by thomasclaudiushuber** [comment]
I don't think that it was not thought through. The property is called "LastChildFill". No matter if the last child is visible, collapsed or hidden, it is still the last child in the DockPanel. That's why the behavior is still correct. 

With the new behavior you're suggesting here, other things pop up. 

Let's assume we would have the behaviour you want and let's look at the sample that I mentioned already above:
``` 
  <DockPanel LastChildFill="True">
        <Button Content="Main menu" DockPanel.Dock="Top"/>
        <Button Content="MainContent 1" />
        <Button Content="MainContent 2" Visibility="Collapsed" />
    </DockPanel>
```
Now let's assume you want to show MainContent 2 instead of MainContent 1

### Step one: Set MainContent 1 to collapsed
Let's do this:
``` 
  <DockPanel LastChildFill="True">
        <Button Content="Main menu" DockPanel.Dock="Top"/>
        <Button Content="MainContent 1" Visibility="Collapsed"/>
        <Button Content="MainContent 2" Visibility="Collapsed" />
    </DockPanel>
```
Now, what will happen here with the behaviour you suggest: The Button fills the leftover space. But in all scenarios I worked at, this is not the behavior I want. The Button should stay where it is. It might be only a short amount of time while the button fills the leftover space, but it could cause a flicker effect.

### Step 2: Set MainContent 1 Visible:
Do it:
``` 
  <DockPanel LastChildFill="True">
        <Button Content="Main menu" DockPanel.Dock="Top"/>
        <Button Content="MainContent 1" Visibility="Collapsed"/>
        <Button Content="MainContent 2" Visibility="`Visible" />
    </DockPanel>
```

So, this problem described here will come up with the new behavior. That's why I still think it is good as it is today, and if you want to do something like you want, you should do it like this:

```
   <DockPanel LastChildFill="True">
        <Button Content="Main menu" DockPanel.Dock="Top"/>
        <Grid>
            <Button Content="MainContent 1" />
            <Button Content="MainContent 2" Visibility="Collapsed" />
        </Grid>
    </DockPanel>
```

**Comment by trodent83** [comment]
@thomasclaudiushuber 
The Problem what you describe there is not really a Problem because you are basicly missusing a list as conditional Content presenter. It is almost as I would say that the ContentPresenter has an incorrect implementation because if I remove the Content, and then add a new one then between the two steps I might get flicker.

**Comment by thomasclaudiushuber** [comment]
@trodent83 This was the only scenario I could think of why you want to have this behavior. What other scenarios do you have why you want the last child fill ignoring collapsed elements?

**Comment by trodent83** [comment]
Any Situation where I have datatemplates which play around with their own visibility.

For example if you have mulltiple pages beside each other, and any of the pages can be set to not visible inside the page, then your current solution would be to basicly make a filtered collectionview which then has to overwatch the IsVisible property of the pages and refilter it self after the change of the property.

**Comment by trodent83** [comment]
@thomasclaudiushuber 
I meant datatemplates, but the situation is the same.
If I will have time then I might make some example,  but I think you are right and we would need other opinions as well.

**Comment by trodent83** [comment]
@thomasclaudiushuber 

So basicly in the following senario it would be quite useful.

Management requirement: On the Right side of the screen you have to have a sidebar which has to fill avaible place completly and the displayed Elements should Change their Looks depending on the avaible place. Also the same functions have to be accessable on the titlebar on the screen Independent of their visibility in the sidebar.

Now the controls mentioned are basicly object delivered from other Departments which are also responsible to provide their or UI designs. The visibility of these controls are usually depend on multiple values and currently you cannot really create a common rule for it.

Now currently the only "reasonable" solution without writing my own Control or changing the dockpanel for this is to provide an interface for all the other departsments where is a you have "IsVisible" property and through this filtering the list on the sidebar. Which works but means basicly that I have to bring display specific Settings in the Viewmodel object, which then won't even effect all the display Areas.

**Comment by t9mike** [comment]
Creating LastVisibleChildFill would allow for backwards compatibility. But the current behavior is something few would want IMHO.
Example of where I would use it below. A number of panels that show/hide based on state. Works for last one, others: not so much.
![image](https://user-images.githubusercontent.com/1008514/111814027-fbc77180-88a7-11eb-86a0-f2d3cf0b7c80.png)
![image](https://user-images.githubusercontent.com/1008514/111814372-611b6280-88a8-11eb-9d15-e974c4a2d55c.png)
